File: Assets/Bundles/Server/Resources/TS/Services/Match/BW/BWService.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local ItemStack = require("Shared/TS/Inventory/ItemStack").ItemStack
local ItemUtil = require("Shared/TS/Item/ItemUtil").ItemUtil
local Network = require("Shared/TS/Network").Network
local SetUtil = require("Shared/TS/Util/SetUtil").SetUtil
local VoxelDataAPI = require("Shared/TS/VoxelWorld/VoxelData/VoxelDataAPI").VoxelDataAPI
local BWService
do
	BWService = setmetatable({}, {
		__tostring = function()
			return "BWService"
		end,
	})
	BWService.__index = BWService
	function BWService.new(...)
		local self = setmetatable({}, BWService)
		return self:constructor(...) or self
	end
	function BWService:constructor(bedService, teamService, matchService, playerService)
		self.bedService = bedService
		self.teamService = teamService
		self.matchService = matchService
		self.playerService = playerService
		self.eliminatedTeams = {}
		self.eliminatedPlayers = {}
		self.winnerDeclared = false
		self.bedHasBeenDestroyed = false
	end
	function BWService:OnStart()
		-- Listen for bed destroy for BW win condition.
		ServerSignals.BedDestroyed:Connect(function()
			self.bedHasBeenDestroyed = true
			self:CheckForWin()
		end)
		-- Listen for entity death for BW win condition, give loot.
		ServerSignals.EntityDeath:ConnectWithPriority(0, function(event)
			if not self.matchService:IsRunning() then
				return nil
			end
			-- Eliminate player, if applicable.
			if TS.instanceof(event.entity, CharacterEntity) then
				if event.entity.player and (event.entity.player:GetTeam() and self.bedService:IsBedDestroyed(event.entity.player:GetTeam().id)) then
					local _eliminatedPlayers = self.eliminatedPlayers
					local _player = event.entity.player
					_eliminatedPlayers[_player] = true
					ServerSignals.PlayerEliminated:Fire({
						player = event.entity.player,
					})
					Network.ServerToClient.PlayerEliminated.Server:FireAllClients(event.entity.player.clientId)
				end
				self:CheckForWin()
			end
			-- Give resources to killer.
			local _result = event.entity
			if _result ~= nil then
				_result = _result.player
			end
			local _condition = _result
			if _condition then
				_condition = TS.instanceof(event.entity, CharacterEntity)
				if _condition then
					local _result_1 = event.killer
					if _result_1 ~= nil then
						_result_1 = _result_1.player
					end
					_condition = _result_1
					if _condition then
						_condition = TS.instanceof(event.killer, CharacterEntity)
					end
				end
			end
			if _condition then
				local deathEntityInv = event.entity:GetInventory()
				local killerEntityInv = event.killer:GetInventory()
				do
					local i = 0
					local _shouldIncrement = false
					while true do
						if _shouldIncrement then
							i += 1
						else
							_shouldIncrement = true
						end
						if not (i < deathEntityInv:GetMaxSlots()) then
							break
						end
						local invSlot = deathEntityInv:GetItem(i)
						if not invSlot then
							continue
						end
						local itemType = invSlot:GetItemType()
						local itemQuantity = invSlot:GetAmount()
						if ItemUtil:IsResource(itemType) then
							killerEntityInv:AddItem(ItemStack.new(itemType, itemQuantity))
						end
					end
				end
			end
		end)
		-- Teammates _cannot_ damage each other.
		ServerSignals.EntityDamage:Connect(function(event)
			local _result = event.fromEntity
			if _result ~= nil then
				_result = _result.player
			end
			local _condition = _result
			if _condition then
				_condition = TS.instanceof(event.fromEntity, CharacterEntity) and (event.entity.player and TS.instanceof(event.entity, CharacterEntity))
			end
			if _condition then
				local fromEntityTeam = event.fromEntity.player:GetTeam()
				local entityTeam = event.entity.player:GetTeam()
				local _result_1 = fromEntityTeam
				if _result_1 ~= nil then
					_result_1 = _result_1.id
				end
				local _result_2 = entityTeam
				if _result_2 ~= nil then
					_result_2 = _result_2.id
				end
				if _result_1 == _result_2 then
					event:SetCancelled(true)
				end
			end
		end)
		-- Teams _cannot_ damage their own beds.
		ServerSignals.BeforeBlockHit:Connect(function(event)
			local teamId = VoxelDataAPI:GetVoxelData(event.BlockPos, "teamId")
			local _condition = teamId ~= nil
			if _condition then
				local _result = event.Player:GetTeam()
				if _result ~= nil then
					_result = _result.id
				end
				_condition = teamId == _result
			end
			if _condition then
				event:SetCancelled(true)
			end
		end)
	end
	function BWService:IsPlayerEliminated(player)
		local _eliminatedPlayers = self.eliminatedPlayers
		local _player = player
		return _eliminatedPlayers[_player] ~= nil
	end
	function BWService:GetEliminatedPlayers()
		return SetUtil:ToArray(self.eliminatedPlayers)
	end
	function BWService:GetEliminatedPlayersOnTeam(team)
		local _exp = SetUtil:ToArray(self.eliminatedPlayers)
		local _arg0 = function(player)
			local _result = player:GetTeam()
			if _result ~= nil then
				_result = _result.id
			end
			return _result == team.id
		end
		-- ▼ ReadonlyArray.filter ▼
		local _newValue = {}
		local _length = 0
		for _k, _v in _exp do
			if _arg0(_v, _k - 1, _exp) == true then
				_length += 1
				_newValue[_length] = _v
			end
		end
		-- ▲ ReadonlyArray.filter ▲
		return _newValue
	end
	function BWService:GetAlivePlayersOnTeam(team)
		local _exp = self.playerService:GetPlayers()
		local _arg0 = function(player)
			local _result = player:GetTeam()
			if _result ~= nil then
				_result = _result.id
			end
			local _condition = _result == team.id
			if _condition then
				local _eliminatedPlayers = self.eliminatedPlayers
				local _player = player
				_condition = not (_eliminatedPlayers[_player] ~= nil)
			end
			return _condition
		end
		-- ▼ ReadonlyArray.filter ▼
		local _newValue = {}
		local _length = 0
		for _k, _v in _exp do
			if _arg0(_v, _k - 1, _exp) == true then
				_length += 1
				_newValue[_length] = _v
			end
		end
		-- ▲ ReadonlyArray.filter ▲
		return _newValue
	end
	function BWService:IsTeamEliminated(team)
		local _eliminatedTeams = self.eliminatedTeams
		local _team = team
		return _eliminatedTeams[_team] ~= nil
	end
	function BWService:CheckForWin()
		local nonEliminatedTeams = {}
		local teams = self.teamService:GetTeams()
		--[[
			* A team is considered eliminated if **all** players
			* are dead AND their bed is destroyed.
		]]
		local _arg0 = function(team)
			local players = SetUtil:ToArray(team:GetPlayers())
			local noPlayersOnTeam = #players == 0
			local _arg0_1 = function(player)
				local _eliminatedPlayers = self.eliminatedPlayers
				local _player = player
				if _eliminatedPlayers[_player] ~= nil then
					return true
				end
				return false
			end
			-- ▼ ReadonlyArray.every ▼
			local _result = true
			for _k, _v in players do
				if not _arg0_1(_v, _k - 1, players) then
					_result = false
					break
				end
			end
			-- ▲ ReadonlyArray.every ▲
			local allPlayersEliminated = _result
			local bedDestroyed = self.bedService:IsBedDestroyed(team.id)
			local isEliminated = noPlayersOnTeam or (allPlayersEliminated and bedDestroyed)
			if not isEliminated then
				local _team = team
				table.insert(nonEliminatedTeams, _team)
			else
				local _eliminatedTeams = self.eliminatedTeams
				local _team = team
				_eliminatedTeams[_team] = true
			end
		end
		for _k, _v in teams do
			_arg0(_v, _k - 1, teams)
		end
		-- If only _one_ team is not eliminated, they win.
		if #nonEliminatedTeams == 1 and self.bedHasBeenDestroyed then
			self:DeclareWinner(nonEliminatedTeams[1])
			return true
		end
		return false
	end
	function BWService:DeclareWinner(team)
		self.winnerDeclared = true
		self.matchService:EndMatch(team)
	end
end
-- (Flamework) BWService metadata
Reflect.defineMetadata(BWService, "identifier", "Bundles/Server/Services/Match/BW/BWService@BWService")
Reflect.defineMetadata(BWService, "flamework:parameters", { "Bundles/Server/Services/Match/BedService@BedService", "Bundles/Server/Services/Global/Team/TeamService@TeamService", "Bundles/Server/Services/Match/MatchService@MatchService", "Bundles/Server/Services/Global/Player/PlayerService@PlayerService" })
Reflect.defineMetadata(BWService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(BWService, "$:flamework@Service", Service, { {} })
return {
	BWService = BWService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Match/BW/BWService.ts
import { CoreServerSignals } from "@Easy/Core/Server/CoreServerSignals";
import { PlayerService } from "@Easy/Core/Server/Services/Player/PlayerService";
import { TeamService } from "@Easy/Core/Server/Services/Team/TeamService";
import { CharacterEntity } from "@Easy/Core/Shared/Entity/Character/CharacterEntity";
import { Game } from "@Easy/Core/Shared/Game";
import { ItemStack } from "@Easy/Core/Shared/Inventory/ItemStack";
import { ItemUtil } from "@Easy/Core/Shared/Item/ItemUtil";
import { Player } from "@Easy/Core/Shared/Player/Player";
import { Team } from "@Easy/Core/Shared/Team/Team";
import { ColorUtil } from "@Easy/Core/Shared/Util/ColorUtil";
import { SetUtil } from "@Easy/Core/Shared/Util/SetUtil";
import { SignalPriority } from "@Easy/Core/Shared/Util/Signal";
import { Theme } from "@Easy/Core/Shared/Util/Theme";
import { OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { Network } from "Shared/Network";
import { BedService } from "../BedService";
import { MatchService } from "../MatchService";

@Service({})
export class BWService implements OnStart {
	constructor(
		private readonly bedService: BedService,
		private readonly teamService: TeamService,
		private readonly matchService: MatchService,
		private readonly playerService: PlayerService,
	) {}

	/** Set of eliminated teams. */
	private eliminatedTeams = new Set<Team>();
	/** Set of eliminated players. */
	private eliminatedPlayers = new Set<Player>();
	/** Whether or not winner has been declared. */
	public winnerDeclared = false;
	/** Whether or not a bed has been destroyed. */
	private bedHasBeenDestroyed = false;

	OnStart(): void {
		// Listen for bed destroy for BW win condition.
		ServerSignals.BedDestroyed.Connect(() => {
			this.bedHasBeenDestroyed = true;
			this.CheckForWin();
		});
		// Listen for entity death for BW win condition, give loot.
		CoreServerSignals.EntityDeath.ConnectWithPriority(SignalPriority.HIGHEST, (event) => {
			if (!this.matchService.IsRunning()) return;
			// Eliminate player, if applicable.
			if (event.entity instanceof CharacterEntity) {
				const team = event.entity.player?.GetTeam();
				if (event.entity.player && team && this.bedService.IsBedDestroyed(team)) {
					this.EliminatePlayer(event.entity.player);
					event.respawnTime = 0;
				}
			}
			// Give resources to killer.
			if (
				event.entity?.player &&
				event.entity instanceof CharacterEntity &&
				event.killer?.player &&
				event.killer instanceof CharacterEntity
			) {
				const deathEntityInv = event.entity.GetInventory();
				const killerEntityInv = event.killer.GetInventory();
				for (let i = 0; i < deathEntityInv.GetMaxSlots(); i++) {
					const invSlot = deathEntityInv.GetItem(i);
					if (!invSlot) continue;
					const itemType = invSlot.GetItemType();
					const itemQuantity = invSlot.GetAmount();
					if (ItemUtil.IsResource(itemType)) {
						killerEntityInv.AddItem(new ItemStack(itemType, itemQuantity));
					}
				}
			}
			// Kill message in chat
			const killerTeam = event.killer?.GetTeam();
			const deadTeam = event.entity.GetTeam();
			if (event.killer && killerTeam && deadTeam) {
				Game.BroadcastMessage(
					ColorUtil.ColoredText(killerTeam.color, event.killer.GetDisplayName()) +
						ColorUtil.ColoredText(Theme.Gray, " killed ") +
						ColorUtil.ColoredText(deadTeam.color, event.entity.GetDisplayName()),
				);
			}
		});
		// Teammates _cannot_ damage each other.
		CoreServerSignals.EntityDamage.Connect((event) => {
			if (event.canDamageAllies) {
				return;
			}

			if (
				event.fromEntity?.player &&
				event.fromEntity instanceof CharacterEntity &&
				event.entity.player &&
				event.entity instanceof CharacterEntity
			) {
				const fromEntityTeam = event.fromEntity.player.GetTeam();
				const entityTeam = event.entity.player.GetTeam();
				if (fromEntityTeam?.id === entityTeam?.id) {
					event.SetCancelled(true);
				}
			}
		});

		// Testing: place block near player ever second
		// SetInterval(1, () => {
		// 	const players = Dependency<PlayerService>().GetPlayers();
		// 	if (players.size() === 0) return;

		// 	const p = players[0];
		// 	if (p.Character) {
		// 		WorldAPI.GetMainWorld().PlaceBlock(
		// 			p.Character.model.transform.position.add(new Vector3(5, 0, 0)),
		// 			ItemType.STONE,
		// 		);
		// 	}
		// });
	}

	public EliminatePlayer(player: Player): void {
		this.eliminatedPlayers.add(player);
		ServerSignals.PlayerEliminated.Fire({ player: player });
		Network.ServerToClient.PlayerEliminated.Server.FireAllClients(player.clientId);
		// this.CheckForWin();
	}

	/**
	 * Checks if a player is on an eliminated team.
	 * @param player A player.
	 * @returns Whether or not `player` is eliminated.
	 */
	public IsPlayerEliminated(player: Player): boolean {
		return this.eliminatedPlayers.has(player);
	}

	/**
	 * Fetch all eliminated players.
	 * @returns A Set of all eliminated players.
	 */
	public GetEliminatedPlayers(): Player[] {
		return SetUtil.ToArray(this.eliminatedPlayers);
	}

	/**
	 * Fetch all eliminated players on a provided team.
	 * @param team A team.
	 * @returns A Set of all eliminated players on `team`.
	 */
	public GetEliminatedPlayersOnTeam(team: Team): Player[] {
		return SetUtil.ToArray(this.eliminatedPlayers).filter((player) => {
			return player.GetTeam()?.id === team.id;
		});
	}

	/**
	 * Fetch all alive players on a provided team.
	 * @param team A team.
	 * @returns A Set of all alive players on `team`.
	 */
	public GetAlivePlayersOnTeam(team: Team): Player[] {
		return this.playerService.GetPlayers().filter((player) => {
			return player.GetTeam()?.id === team.id && !this.eliminatedPlayers.has(player);
		});
	}

	/**
	 * Checks if a team is eliminated.
	 * @param team A team.
	 * @returns Whether or not `team` is eliminated.
	 */
	public IsTeamEliminated(team: Team): boolean {
		return this.eliminatedTeams.has(team);
	}

	/** Check for win condition. Returns whether or not win condition was met.*/
	private CheckForWin(): boolean {
		const nonEliminatedTeams: Team[] = [];
		const teams = this.teamService.GetTeams();
		/*
		 * A team is considered eliminated if **all** players
		 * are dead AND their bed is destroyed.
		 */
		teams.forEach((team) => {
			const players = SetUtil.ToArray(team.GetPlayers());
			const noPlayersOnTeam = players.size() === 0;
			const allPlayersEliminated = players.every((player) => {
				if (this.eliminatedPlayers.has(player)) return true;
				return false;
			});
			const bedDestroyed = this.bedService.IsBedDestroyed(team);
			const isEliminated = noPlayersOnTeam || (allPlayersEliminated && bedDestroyed);
			if (!isEliminated) {
				nonEliminatedTeams.push(team);
			} else {
				this.eliminatedTeams.add(team);
			}
		});
		/* If only _one_ team is not eliminated, they win. */
		if (nonEliminatedTeams.size() === 1 && this.bedHasBeenDestroyed) {
			this.DeclareWinner(nonEliminatedTeams[0]);
			return true;
		}
		return false;
	}

	/** Declares a team as winner. */
	private DeclareWinner(team: Team) {
		this.winnerDeclared = true;
		this.matchService.EndMatch(team);
	}
}
